File: Problem46.java
//Time Complexity: O(n^2)
//Space Complexity: O(n*n)with recursive stack
//Did it run on leetcode: yes
//Problems faced any: No

import java.util.ArrayList;
import java.util.List;

public class Problem46 {
    private List<List<Integer>> output = new ArrayList<>();

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<Integer> list = new ArrayList<>();
        findSumII(root, 0, sum, list);
        return output;

    }

    void findSumII(TreeNode node, int prev, int sum, List<Integer> list){
        if(node==null)
            return;

        //on each recursive we keep adding sum of the path
        int currSum = prev + node.val;

        //we create a new list and add all the node values of the path
        List<Integer> newList = new ArrayList<>();
        newList.addAll(list);
        newList.add(node.val);

        if(node.left == null & node.right==null){
            //if the path sum is equal to given sum we will add it to the output list
            if(currSum == sum){
                output.add(newList);
                return;
            }else
                return;
        }

        findSumII(node.left, currSum, sum, newList);
        findSumII(node.right, currSum, sum, newList);
    }
}

File: Problem-46.py
# leetcode: 113
# Time Complexity - O(N)
# Space complexity -O(N)



# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        res=[]
        path=[]
        self.dfs(root,sum,res,path)
        return res
    
    def dfs(self,root,sum,res,path):
        if not root:
            return
        else:
            path.append(root.val)
        
        if (root.left == None and root.right == None and sum == root.val): 
            res.append(list(path))
            return
        
        if root.left:
            self.dfs(root.left,sum-root.val,res,path)
            path.pop()
        if root.right:
            self.dfs(root.right,sum-root.val,res,path)
            path.pop()
    